Resolution of ObjectCreationExprs broken 3.26.2

Hi,

this is the other problem I mentioned in #4557. Since 3.26.2 the resolution of ObjectCreationExprs is broken. I don't know exactly which change triggers this, but the test (see below) runs without problems with tag 3.26.1 but with tag 3.26.2 the ObjectCreationExpr resolution fails. I have tried to find the change causing this, but without success. There were a lot of changes between the two tags and, above all, none that I could blame at first glance. I am also not sure if this is the only expression that cannot be resolved.

<details>
  <summary>Test</summary>

```Java
package com.github.javaparser.symbolsolver;

import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.Expression;
import com.github.javaparser.ast.expr.NameExpr;
import com.github.javaparser.ast.expr.ObjectCreationExpr;
import com.github.javaparser.resolution.declarations.ResolvedConstructorDeclaration;
import com.github.javaparser.resolution.declarations.ResolvedValueDeclaration;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

import java.util.List;

public class Test {

    @org.junit.jupiter.api.Test
    public void test() {
        String code = "public class Board {\n" +
                      "\n" +
                      "    class Field {\n" +
                      "        private final Board board;\n" +
                      "        private final int x;\n" +
                      "        private final int y;\n" +
                      "\n" +
                      "        public Field(Board board, int x, int y) {\n" +
                      "            this.board = board;\n" +
                      "            this.x = x;\n" +
                      "            this.y = y;\n" +
                      "        }\n" +
                      "    }\n" +
                      "\n" +
                      "    /**\n" +
                      "     * Edge size of the Sudoku board.\n" +
                      "     */\n" +
                      "    public static final int SIZE = 9;\n" +
                      "    private final Field[] board = new Field[SIZE * SIZE];\n" +
                      "\n" +
                      "    /**\n" +
                      "     * Creates a Sudoku board with all of its fields being empty.\n" +
                      "     */\n" +
                      "    public Board() {\n" +
                      "        for (int x = 0; x < SIZE; x++)\n" +
                      "            for (int y = 0; y < SIZE; y++)\n" +
                      "                board[SIZE * y + x] = new Field(this, x, y);\n" +
                      "    }\n" +
                      "}";
        StaticJavaParser.getParserConfiguration().setSymbolResolver(new JavaSymbolSolver(new ReflectionTypeSolver()));
        CompilationUnit cu = StaticJavaParser.parse(code);

        final List<ObjectCreationExpr> objectCreationExprs = cu.findAll(ObjectCreationExpr.class);

        for (ObjectCreationExpr objectCreationExpr : objectCreationExprs) {
            System.err.println(objectCreationExpr.toString());
            for (Expression e : objectCreationExpr.getArguments()) {
                System.err.println(e.toString());
                if (e.isNameExpr()) {
                    NameExpr ne = e.asNameExpr();
                    final ResolvedValueDeclaration resolve = ne.resolve();
                    System.err.println(resolve);
                    System.err.println(resolve.getType().describe());
                }
            }
            final ResolvedConstructorDeclaration resolve = objectCreationExpr.resolve();
            System.err.println(resolve.getQualifiedName());
            System.err.println(resolve.getQualifiedSignature());
        }
    }
}
```

</details>

Output 3.26.1
```
[INFO] Running com.github.javaparser.symbolsolver.Test
new Field(this, x, y)
this
x
JavaParserVariableDeclaration{x}
int
y
JavaParserVariableDeclaration{y}
int
Board.Field.Field
Board.Field.Field(Board, int, int)
```

Output 3.26.2
```
[INFO] Running com.github.javaparser.symbolsolver.Test
new Field(this, x, y)
this
x
JavaParserVariableDeclaration{x}
int
y
JavaParserVariableDeclaration{y}
int
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.326 s <<< FAILURE! -- in com.github.javaparser.symbolsolver.Test
[ERROR] com.github.javaparser.symbolsolver.Test.test -- Time elapsed: 0.302 s <<< ERROR!
UnsolvedSymbolException{context='x', name='Solving x', cause='null'}
        at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:339)
        at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:63)
        at com.github.javaparser.ast.expr.NameExpr.accept(NameExpr.java:80)
        at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:533)
        at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:394)
        at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:273)
        at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:231)
        at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:141)
        at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:202)
        at com.github.javaparser.ast.expr.ObjectCreationExpr.resolve(ObjectCreationExpr.java:379)
        at com.github.javaparser.symbolsolver.Test.test(Test.java:63)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
```

I'm happy to provide more information and try to help with the solution, but I can't get any further for now.

best regards,
Kim

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/JavaParserFactory.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/JavaParserFactory.java
@@ -23,8 +23,6 @@ package com.github.javaparser.symbolsolver.javaparsermodel;
 
 import static com.github.javaparser.resolution.Navigator.demandParentNode;
 
-import java.util.Optional;
-
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.Node;
 import com.github.javaparser.ast.body.*;
@@ -36,6 +34,7 @@ import com.github.javaparser.resolution.SymbolDeclarator;
 import com.github.javaparser.resolution.TypeSolver;
 import com.github.javaparser.symbolsolver.javaparsermodel.contexts.*;
 import com.github.javaparser.symbolsolver.javaparsermodel.declarators.*;
+import java.util.Optional;
 
 /**
  * @author Federico Tomassetti
@@ -186,7 +185,7 @@ public class JavaParserFactory {
         }
         if (node instanceof ExpressionStmt) {
             ExpressionStmt expressionStmt = (ExpressionStmt) node;
-            if (expressionStmt.getExpression().isVariableDeclarationExpr()){
+            if (expressionStmt.getExpression().isVariableDeclarationExpr()) {
                 return new VariableSymbolDeclarator(
                         expressionStmt.getExpression().asVariableDeclarationExpr(), typeSolver);
             }
@@ -199,11 +198,11 @@ public class JavaParserFactory {
         if (node instanceof ForStmt) {
             ForStmt forStmt = (ForStmt) node;
             Optional<VariableDeclarationExpr> variableDecl = forStmt.getInitialization().stream()
-            		.filter(expr -> expr.isVariableDeclarationExpr())
-            		.map(expr -> expr.asVariableDeclarationExpr())
-            		.findFirst();
+                    .filter(expr -> expr.isVariableDeclarationExpr())
+                    .map(expr -> expr.asVariableDeclarationExpr())
+                    .findFirst();
             if (variableDecl.isPresent()) {
-            	return new VariableSymbolDeclarator(variableDecl.get(), typeSolver);
+                return new VariableSymbolDeclarator(variableDecl.get(), typeSolver);
             }
         }
         return new NoSymbolDeclarator<>(node, typeSolver);

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/JavaParserFactory.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/JavaParserFactory.java
@@ -23,6 +23,8 @@ package com.github.javaparser.symbolsolver.javaparsermodel;
 
 import static com.github.javaparser.resolution.Navigator.demandParentNode;
 
+import java.util.Optional;
+
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.Node;
 import com.github.javaparser.ast.body.*;
@@ -184,15 +186,25 @@ public class JavaParserFactory {
         }
         if (node instanceof ExpressionStmt) {
             ExpressionStmt expressionStmt = (ExpressionStmt) node;
-            if (expressionStmt.getExpression() instanceof VariableDeclarationExpr) {
+            if (expressionStmt.getExpression().isVariableDeclarationExpr()){
                 return new VariableSymbolDeclarator(
-                        (VariableDeclarationExpr) (expressionStmt.getExpression()), typeSolver);
+                        expressionStmt.getExpression().asVariableDeclarationExpr(), typeSolver);
             }
             return new NoSymbolDeclarator<>(expressionStmt, typeSolver);
         }
         if (node instanceof ForEachStmt) {
             ForEachStmt foreachStmt = (ForEachStmt) node;
             return new VariableSymbolDeclarator(foreachStmt.getVariable(), typeSolver);
+        }
+        if (node instanceof ForStmt) {
+            ForStmt forStmt = (ForStmt) node;
+            Optional<VariableDeclarationExpr> variableDecl = forStmt.getInitialization().stream()
+            		.filter(expr -> expr.isVariableDeclarationExpr())
+            		.map(expr -> expr.asVariableDeclarationExpr())
+            		.findFirst();
+            if (variableDecl.isPresent()) {
+            	return new VariableSymbolDeclarator(variableDecl.get(), typeSolver);
+            }
         }
         return new NoSymbolDeclarator<>(node, typeSolver);
     }

--- a/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/TypeExtractor.java
+++ b/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/TypeExtractor.java
@@ -538,7 +538,7 @@ public class TypeExtractor extends DefaultVisitorAdapter {
             return result;
         }
         if (parentNode instanceof ObjectCreationExpr) {
-        	ObjectCreationExpr expr = (ObjectCreationExpr) parentNode;
+            ObjectCreationExpr expr = (ObjectCreationExpr) parentNode;
             ResolvedType result = expr.getType().resolve();
 
             if (solveLambdas) {

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4560Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4560Test.java
@@ -20,15 +20,14 @@
 
 package com.github.javaparser.symbolsolver;
 
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+
 import com.github.javaparser.JavaParser;
 import com.github.javaparser.JavaParserAdapter;
 import com.github.javaparser.ParserConfiguration;
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.expr.MethodCallExpr;
 import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
-
-import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
-
 import org.junit.jupiter.api.Test;
 
 public class Issue4560Test extends AbstractResolutionTest {
@@ -36,14 +35,14 @@ public class Issue4560Test extends AbstractResolutionTest {
     @Test
     void test() {
 
-    	String code = "public class MyExample {\n"
-				+ "\n"
-				+ "  public static void main(String[] args) {\n"
-				+ "    \"\"\"\n"
-				+ "        Hello multiple %s!\n"
-				+ "        \"\"\".format(\"world\");\n"
-				+ "  }\n"
-				+ "}";
+        String code = "public class MyExample {\n"
+                + "\n"
+                + "  public static void main(String[] args) {\n"
+                + "    \"\"\"\n"
+                + "        Hello multiple %s!\n"
+                + "        \"\"\".format(\"world\");\n"
+                + "  }\n"
+                + "}";
 
         ParserConfiguration parserConfiguration = new ParserConfiguration()
                 .setLanguageLevel(ParserConfiguration.LanguageLevel.JAVA_15)

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4568Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4568Test.java
@@ -20,6 +20,8 @@
 
 package com.github.javaparser.symbolsolver;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
 import com.github.javaparser.JavaParser;
 import com.github.javaparser.JavaParserAdapter;
 import com.github.javaparser.ParserConfiguration;
@@ -30,8 +32,6 @@ import com.github.javaparser.ast.expr.ObjectCreationExpr;
 import com.github.javaparser.resolution.declarations.ResolvedConstructorDeclaration;
 import com.github.javaparser.resolution.declarations.ResolvedValueDeclaration;
 import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-
 import org.junit.jupiter.api.Test;
 
 public class Issue4568Test extends AbstractResolutionTest {
@@ -39,44 +39,44 @@ public class Issue4568Test extends AbstractResolutionTest {
     @Test
     void test() {
 
-		String code = "public class Board {\n" +
-                "\n" +
-                "    class Field {\n" +
-                "        private final Board board;\n" +
-                "        private final int x;\n" +
-                "        private final int y;\n" +
-                "\n" +
-                "        public Field(Board board, int x, int y) {\n" +
-                "            this.board = board;\n" +
-                "            this.x = x;\n" +
-                "            this.y = y;\n" +
-                "        }\n" +
-                "    }\n" +
-                "\n" +
-                "    public static final int SIZE = 9;\n" +
-                "    private final Field[] board;\n" +
-                "\n" +
-                "    public Board() {\n" +
-                "        for (int x = 0; x < SIZE; x++)\n" +
-                "            for (int y = 0; y < SIZE; y++)\n" +
-                "                board[SIZE * y + x] = new Field(this, x, y);\n" +
-                "    }\n" +
-                "}";
-		ParserConfiguration parserConfiguration = new ParserConfiguration()
-                .setSymbolResolver(symbolResolver(defaultTypeSolver()));
+        String code = "public class Board {\n" + "\n"
+                + "    class Field {\n"
+                + "        private final Board board;\n"
+                + "        private final int x;\n"
+                + "        private final int y;\n"
+                + "\n"
+                + "        public Field(Board board, int x, int y) {\n"
+                + "            this.board = board;\n"
+                + "            this.x = x;\n"
+                + "            this.y = y;\n"
+                + "        }\n"
+                + "    }\n"
+                + "\n"
+                + "    public static final int SIZE = 9;\n"
+                + "    private final Field[] board;\n"
+                + "\n"
+                + "    public Board() {\n"
+                + "        for (int x = 0; x < SIZE; x++)\n"
+                + "            for (int y = 0; y < SIZE; y++)\n"
+                + "                board[SIZE * y + x] = new Field(this, x, y);\n"
+                + "    }\n"
+                + "}";
+        ParserConfiguration parserConfiguration =
+                new ParserConfiguration().setSymbolResolver(symbolResolver(defaultTypeSolver()));
 
-        CompilationUnit cu = JavaParserAdapter.of(new JavaParser(parserConfiguration)).parse(code);
+        CompilationUnit cu =
+                JavaParserAdapter.of(new JavaParser(parserConfiguration)).parse(code);
 
-		ObjectCreationExpr oce = cu.findFirst(ObjectCreationExpr.class).get();
+        ObjectCreationExpr oce = cu.findFirst(ObjectCreationExpr.class).get();
 
-		for (Expression e : oce.getArguments()) {
-			if (e.isNameExpr()) {
-				NameExpr ne = e.asNameExpr();
-				final ResolvedValueDeclaration rvd = ne.resolve();
-				assertEquals("int", rvd.getType().describe());
-			}
-		}
-		final ResolvedConstructorDeclaration rcd = oce.resolve();
-		assertEquals("Board.Field.Field(Board, int, int)", rcd.getQualifiedSignature());
-	}
+        for (Expression e : oce.getArguments()) {
+            if (e.isNameExpr()) {
+                NameExpr ne = e.asNameExpr();
+                final ResolvedValueDeclaration rvd = ne.resolve();
+                assertEquals("int", rvd.getType().describe());
+            }
+        }
+        final ResolvedConstructorDeclaration rcd = oce.resolve();
+        assertEquals("Board.Field.Field(Board, int, int)", rcd.getQualifiedSignature());
+    }
 }

--- a/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4568Test.java
+++ b/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue4568Test.java
@@ -1,0 +1,82 @@
+/*
+ * Copyright (C) 2013-2024 The JavaParser Team.
+ *
+ * This file is part of JavaParser.
+ *
+ * JavaParser can be used either under the terms of
+ * a) the GNU Lesser General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ * b) the terms of the Apache License
+ *
+ * You should have received a copy of both licenses in LICENCE.LGPL and
+ * LICENCE.APACHE. Please refer to those files for details.
+ *
+ * JavaParser is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+package com.github.javaparser.symbolsolver;
+
+import com.github.javaparser.JavaParser;
+import com.github.javaparser.JavaParserAdapter;
+import com.github.javaparser.ParserConfiguration;
+import com.github.javaparser.ast.CompilationUnit;
+import com.github.javaparser.ast.expr.Expression;
+import com.github.javaparser.ast.expr.NameExpr;
+import com.github.javaparser.ast.expr.ObjectCreationExpr;
+import com.github.javaparser.resolution.declarations.ResolvedConstructorDeclaration;
+import com.github.javaparser.resolution.declarations.ResolvedValueDeclaration;
+import com.github.javaparser.symbolsolver.resolution.AbstractResolutionTest;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public class Issue4568Test extends AbstractResolutionTest {
+
+    @Test
+    void test() {
+
+		String code = "public class Board {\n" +
+                "\n" +
+                "    class Field {\n" +
+                "        private final Board board;\n" +
+                "        private final int x;\n" +
+                "        private final int y;\n" +
+                "\n" +
+                "        public Field(Board board, int x, int y) {\n" +
+                "            this.board = board;\n" +
+                "            this.x = x;\n" +
+                "            this.y = y;\n" +
+                "        }\n" +
+                "    }\n" +
+                "\n" +
+                "    public static final int SIZE = 9;\n" +
+                "    private final Field[] board;\n" +
+                "\n" +
+                "    public Board() {\n" +
+                "        for (int x = 0; x < SIZE; x++)\n" +
+                "            for (int y = 0; y < SIZE; y++)\n" +
+                "                board[SIZE * y + x] = new Field(this, x, y);\n" +
+                "    }\n" +
+                "}";
+		ParserConfiguration parserConfiguration = new ParserConfiguration()
+                .setSymbolResolver(symbolResolver(defaultTypeSolver()));
+
+        CompilationUnit cu = JavaParserAdapter.of(new JavaParser(parserConfiguration)).parse(code);
+
+		ObjectCreationExpr oce = cu.findFirst(ObjectCreationExpr.class).get();
+
+		for (Expression e : oce.getArguments()) {
+			if (e.isNameExpr()) {
+				NameExpr ne = e.asNameExpr();
+				final ResolvedValueDeclaration rvd = ne.resolve();
+				assertEquals("int", rvd.getType().describe());
+			}
+		}
+		final ResolvedConstructorDeclaration rcd = oce.resolve();
+		assertEquals("Board.Field.Field(Board, int, int)", rcd.getQualifiedSignature());
+	}
+}

